[GeoMechanicsApplication] Add Thermal Element

mnabideltares opened the pull request:
**📝 Description**
This PR is continiouation of : https://github.com/KratosMultiphysics/Kratos/pull/10616
The former PR is stopped development and shiftet to this branch because we needed a branch which can be operated on our local testbench (Teamcity) and SonarCloud, namely starts with "geo/..."

All the implementations of the old branch are already implemented in this branch and continued with further development.
In this branch we added a transient thermal element and its required solver(s). Neumann BC is also added.

This branch has already 3 sub-branches:
geo/10372-HeatTransfer-temp-branch   => for MicroClimate BC
geo/10372-HeatTransfer-temp-branch-cosim => coupling pressure and thermal elements by CoSim
geo/10372-HeatTransfer-temp-branch-coupling => coupling pressure and thermal elements by python solver

**🆕 Changelog**
[1] Define thermal element with element classes
[2] Add the input variables needed for heat
[3] Defind solver
[4] Define strategy
[5] Add dispersion matrix/law.
[6] Build matrices, stiffness, right hand side, mass ...
[7] Add boundary conditions (Neumann)
[10] Give output for GID
[11] Fix bugs

mnabideltares committed:
issue: 10372 | Task 1, Added thermal element with element class (empty classes)

mnabideltares committed:
issue: 10372 | Task 2, Added material properties needed for heat.

mnabideltares committed:
issue: 10372 | Task 3, Added solver for heat (python).

mnabideltares committed:
issue: 10372 | Task 4, Added schemes for strategy.

(1) Inheritated from U_Pw which is not relevant. (2) UpdateVariablesDerivatives is not defined (need to be removed?)

mnabideltares committed:
issue: 10372 | Task 4A, Added schemes for strategy.

Continuation

mnabideltares committed:
issue: 10372 | Task 2, Added dispersion 2D law for thermal element

mnabideltares committed:
issue: 10372 | Task 6, Built matrices for thermal element

mnabideltares committed:
Revert "issue: 10372 | Task 6, Built matrices for thermal element"

This reverts commit 137f0c00a82c1a0f7883c62218e67f29fd6aa70e.

mnabideltares committed:
Revert "issue: 10372 | Task 2, Added dispersion 2D law for thermal element"

This reverts commit 346200f1f1cdd902452de8f90a207aa4b0dd2a6b.

mnabideltares committed:
Revert "issue: 10372 | Task 4A, Added schemes for strategy."

This reverts commit 70dcedb27be48271fe33bc5fd7c40f4d6d733891.

mnabideltares committed:
Revert "issue: 10372 | Task 4, Added schemes for strategy."

This reverts commit a71586d09fe99c03b7d40176a2578a4d8f340974.

mnabideltares committed:
Revert "issue: 10372 | Task 3, Added solver for heat (python)."

This reverts commit 122537b40e02d1cad4a468353a775919a4c6312e.

mnabideltares committed:
Revert "issue: 10372 | Task 2, Added material properties needed for heat."

This reverts commit 72e2475b49fd7015bab8f0a0f67eb8255be4a196.

mnabideltares committed:
Revert "issue: 10372 | Task 1, Added thermal element with element class (empty classes)"

This reverts commit ef8d06bf20f5b2f43dfea3973dfec2355fb72f74.

mnabideltares committed:
Merge branch 'master' into Geo-HeatTransfer-10372

mnabideltares committed:
Tast 1: Define thermal element with element classes (empty)

mnabideltares committed:
Task 2: Add the input variables needed for heat

NB. Variables for the heat flux bounday are also added.

mnabideltares committed:
Task 3: Defind solver

NB. The solver is added as py file(s)

mnabideltares committed:
Task 4: Define strategy

mnabideltares committed:
Task 5: Add dispersion matrix/law.

mnabideltares committed:
Task 6: Build matrices, stiffness, right hand side, mass ...

NB. strategy is added to python, which is a part of Task 4, but was forgotten.

mnabideltares committed:
Task 7: Add boundary conditions (Neumann)

mnabideltares committed:
Task 10: Give output for GID

mnabideltares committed:
Task 11: Dix bugs

mnabideltares committed:
Task 11: Fix bugs

mnabideltares committed:
Task 11: Fix bug

backward_euler_quasistatic_T_scheme.hpp is automatically renamed by capital letters, and it gives issues under Linux. Here we delete it, and next commit add in the form of small letters.

mnabideltares committed:
Task 11: Fix bug

following by previous commit to add the deleted file back.

mnabideltares committed:
Task 11: Fix bug

1) override is used.  2) mDeltaTime is read from parent class.

mnabideltares committed:
Modifications based on reviewer requests

mnabideltares committed:
Task 11: Removed unused variables

mnabideltares committed:
Modifications based on reviewer (Vahid-Galavi) requests

mnabideltares committed:
another modification based on reviewer

mnabideltares committed:
Fixes based on review

mnabideltares committed:
Three test cases are added for the thermal (temperature) problem in Geo

mnabideltares committed:
fixed a small bug related to added test cases

mnabideltares committed:
Merge remote-tracking branch 'origin/master' into geo/10372-HeatTransfer-temp

# Conflicts:
#	applications/GeoMechanicsApplication/custom_python/add_custom_strategies_to_python.cpp
#	applications/GeoMechanicsApplication/geo_mechanics_application.cpp
#	applications/GeoMechanicsApplication/geo_mechanics_application.h
#	applications/GeoMechanicsApplication/tests/test_GeoMechanicsApplication.py

mnabideltares committed:
code smells are fixed

mnabideltares committed:
minor fix

mnabideltares committed:
Merge branch 'master' into geo/10372-HeatTransfer-temp

# Conflicts:
#	applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py
#	applications/GeoMechanicsApplication/tests/test_GeoMechanicsApplication.py

Also "Node<3>" is replaced to "Node" in new elements and conditions of thermal

mnabideltares committed:
Thermal solver is decoupled from U-Pw

Node definition was mistekenly not stagged in last commit. Here, committed again

huhaas commented on applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp:
```diff
@@ -52,9 +52,12 @@ PYBIND11_MODULE(KratosGeoMechanicsApplication,m)
     //Registering variables in python
     KRATOS_REGISTER_IN_PYTHON_VARIABLE( m, VELOCITY_COEFFICIENT )
     KRATOS_REGISTER_IN_PYTHON_VARIABLE( m, DT_PRESSURE_COEFFICIENT )
+	KRATOS_REGISTER_IN_PYTHON_VARIABLE( m, DT_TEMPERATURE_COEFFICIENT )
```
Please make this line consistent with others.

mnabideltares replied:
@huhaas, Thank you for the review,

The line is made consistent.

huhaas commented on applications/GeoMechanicsApplication/custom_strategies/schemes/backward_euler_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,79 @@
+    using TSystemMatrixType = typename BaseType::TSystemMatrixType;
+    using TSystemVectorType = typename BaseType::TSystemVectorType;
+    using LocalSystemVectorType = typename BaseType::LocalSystemVectorType;
+    using LocalSystemMatrixType = typename BaseType::LocalSystemMatrixType;
```
Not all of these are used. Please remove those.

mnabideltares replied:
Removed the unused lines

huhaas commented on applications/GeoMechanicsApplication/custom_strategies/schemes/backward_euler_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,79 @@
+
+protected:
+
+    /// Member Variables
```
Incorrect comment. All 3 lines above the functions are not providing more clarity. In my opinion all 3 comment blocks can be removed.

mnabideltares replied:
Removed Line 60

huhaas commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -32,6 +32,11 @@ class GeoFlowApplyConstantScalarValueProcess : public Kratos::ApplyConstantScala
         return mvariable_name == "WATER_PRESSURE";
     }
 
+    bool hasTemperature()
```
I don't think the D-Geo-Flow application uses (or wants to) Temperature. This addition mixes research with production.

mnabideltares replied:
The temperature check is removed from dgeoflow

huhaas commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -295,6 +311,8 @@ void KratosGeoMechanicsApplication::Register() {
 
     KRATOS_REGISTER_CONSTITUTIVE_LAW("LinearElastic2DBeamLaw",        mLinearElastic2DBeamLaw)
 
+	KRATOS_REGISTER_CONSTITUTIVE_LAW("GeoThermalDispersion2DLaw", mGeoThermalDispersion2DLaw)
```
Tab instead of spaces

mnabideltares replied:
Done

huhaas commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_T_solver.py:
```diff
@@ -0,0 +1,210 @@
+            "number_cycles"              : 5,
+            "increase_factor"            : 2.0,
+            "reduction_factor"           : 0.5,
+            "realised_factor"            : 1.0,
```
`realised_factor` is removed PR [link](https://github.com/KratosMultiphysics/Kratos/pull/11149). Please remove it here.

mnabideltares replied:
Done

huhaas commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py:
```diff
@@ -22,10 +22,16 @@ def CreateSolver(model, custom_settings):
             custom_settings["solver_settings"]["time_stepping"].AddValue("end_time", custom_settings["problem_data"]["end_time"])
             solver_module_name = "geomechanics_Pw_solver"
+
+        elif (solver_type.lower() == "t" or solver_type.lower() == "geomechanics_t_solver" or solver_type.lower() == "twophase"):
```
The style of this if-tree changed by moving the `.lower()` to line 17. Please update accordingly.

mnabideltares replied:
The stlye is changed consistent with the rest of variables, like Pw

huhaas commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py:
```diff
@@ -22,10 +22,16 @@ def CreateSolver(model, custom_settings):
         else:
             err_msg =  "The requested solver type \"" + solver_type + "\" is not in the python solvers wrapper\n"
-            err_msg += "Available options are: \"geomechanics_U_Pw_solver\", \"geomechanics_Pw_solver\""
+            err_msg += "Available options are: \"geomechanics_U_Pw_solver\""
```
Do we now support less solvers?

mnabideltares replied:
geomechanics_Pw_solver and  geomechanics_T_solver are added to the error message

huhaas commented on applications/GeoMechanicsApplication/tests/test_thermal_fixed_temperature_2D3N.gid/ProjectParameters.json:
```diff
@@ -0,0 +1,146 @@
+        "number_cycles":                      100,
+        "reduction_factor":                   0.5,
+        "increase_factor":                    2.0,
+        "realised_factor":                    0.9999,
```
See comment above.

mnabideltares replied:
increase_factor  is removed

huhaas commented on applications/GeoMechanicsApplication/tests/test_transient_thermal.py:
```diff
@@ -0,0 +1,46 @@
+        # Code here will be placed AFTER every test in this TestCase.
+        pass
+
+    def test_thermal_fixed_temperature_2D3N(self):
```
Judging from the size of this PR I expected a bit more tests than 3.

mnabideltares replied:
New test case(s) will be added soon.

huhaas commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,180 @@
+    const Matrix& Jacobian,
+    const double& Weight)
+{
+    double NormalVector[3];
```
Improvement for later. This function body is present in multiple files. Should be fairly ease to move it to a common section and remove all the template specialization.

mnabideltares replied:
Will checked and improved later

huhaas commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp:
```diff
@@ -0,0 +1,215 @@
+ * @ingroup GeoMechanicsApplication
+ * @brief This class defines the thermal dispersion for heat cases
+ * @details This class derives from the linear elastic case on 3D
+ * @author Vahid Galavi
```
Copy-paste error?

mnabideltares replied:
Removed

huhaas commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp:
```diff
@@ -0,0 +1,215 @@
+ * @brief It calculates the constitutive matrix C
+ * @param C The constitutive matrix
+ */
+    //void CalculateThermalDispersionMatrix(Matrix& C, ConstitutiveLaw::Parameters& rValues);
```
Please remove this :)

mnabideltares replied:
line 147 is uncommented, it is necessary.

huhaas commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp:
```diff
@@ -0,0 +1,215 @@
+    //  * @param rValues The internal values of the law
+    //  * @param rStrainVector The strain vector in Voigt notation
+    //  */
+    // void CalculateCauchyGreenStrain(ConstitutiveLaw::Parameters& rValues, Vector& rStrainVector) override;
```
Function does not seem to exist.

huhaas commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp:
```diff
@@ -0,0 +1,215 @@
+ * @details This class derives from the linear elastic case on 3D
+ * @author Vahid Galavi
+ */
+class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoThermalDispersion2DLaw
```
Is it correct to inherit from something with K0?

mnabideltares replied:
The inherittance is changed to ConstitutiveLaw

huhaas commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+
+    	const unsigned int N_DOF = this->GetNumberOfDOF();
+
+        if (rMassMatrix.size1() != N_DOF)
```
Can it be assumed that `rMassMatrix` always square?

mnabideltares replied:
Yes, it is always a square matrix

huhaas commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_T_solver.py:
```diff
@@ -0,0 +1,210 @@
+                KratosMultiphysics.Logger.PrintInfo("GeoMechanics_T_Solver, scheme", "Backward Euler Steady-state heat transfer.")
+                scheme = scheme = KratosGeo.BackwardEulerQuasistaticTScheme()
+        else:
+            raise Exception("Apart from Newmark, other scheme_type are not available.")
```
What about backward_euler?

mnabideltares replied:
backward_euler is added to the message of Exception

mnabideltares committed:
Chanes Based on review from huhaas

mnabideltares committed:
Merge branch 'master' into geo/10372-HeatTransfer-temp

# Conflicts:
#	applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp
#	applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h

mnabideltares committed:
A small fix

mnabideltares committed:
realised_factor is removed from the test cases

mnabideltares committed:
Modified CalculateIntegrationCoefficient by generalizing it by tempelate

mnabideltares committed:
Thermal element is made possible to use more gauss points (applied for high order elements)

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,130 @@
+        const double dA = std::sqrt(NormalVector[0] * NormalVector[0]
+            + NormalVector[1] * NormalVector[1]
+            + NormalVector[2] * NormalVector[2]);
+        rIntegrationCoefficient = dA * Weight;
```
Please reuse the implementations of Euclidian norm and cross product from Kratos\utilities\math_utils.h

mnabideltares replied:
Vector NormalVector = ZeroVector(TDim);

    if (TDim == 2)  {
        NormalVector = column(Jacobian, 0);
    }
    else if (TDim == 3)   {
        MathUtils<double>::CrossProduct(NormalVector, column(Jacobian, 0), column(Jacobian, 1));
    }
    rIntegrationCoefficient = Weight * MathUtils<double>::Norm(NormalVector);

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.hpp:
```diff
@@ -0,0 +1,125 @@
+
+    GeometryData::IntegrationMethod mThisIntegrationMethod;
+    static constexpr unsigned int nDof = 1;
+    static constexpr unsigned int conditionSize = TNumNodes * nDof;
```
Is the convention for member variables naming these as nNDof and mConditionSize?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.hpp:
```diff
@@ -0,0 +1,115 @@
+        array_1d<double,TNumNodes> TVector;
+    };
+    
+    // Member Variables
```
comment without implementation

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.hpp:
```diff
@@ -0,0 +1,115 @@
+
+private:
+    
+    // Member Variables
```
comment without implementation

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp:
```diff
@@ -0,0 +1,75 @@
+        C(INDEX_2D_HEAT_X, INDEX_2D_HEAT_X) = c1 * thermalCondSolidXX + c0;
+        C(INDEX_2D_HEAT_X, INDEX_2D_HEAT_Y) = c1 * thermalCondSolidXY;
+        C(INDEX_2D_HEAT_Y, INDEX_2D_HEAT_X) = c1 * thermalCondSolidYX;
+        C(INDEX_2D_HEAT_Y, INDEX_2D_HEAT_Y) = c1 * thermalCondSolidYY + c0;
```
proposed rename c0 --> cWater c1 --> cSoil or cSolid or cSkeleton and leave thermalCondWater outside the definition of c1.
readability is then gained: C = cWater * thermalCondWater + cSolid * thermalCondSolid

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+
+    	const GeometryType& rGeom = this->GetGeometry();
+        const unsigned int N_DOF = this->GetNumberOfDOF();
+        unsigned int index = 0;
```
Why the extra variable index, rElementalDofList[i] = rGeom[i].pGetDof(TEMPERATURE)); is clearer to me.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+        if (rElementalDofList.size() != N_DOF)
+            rElementalDofList.resize(N_DOF);
+
+        for (unsigned int i = 0; i < TNumNodes; ++i)
```
The implicit knowledge is that N_DOF == TNumNodes.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+    // ============================================================================================
+    // ============================================================================================
+    template<unsigned int TDim, unsigned int TNumNodes>
+    void TransientThermalElement<TDim, TNumNodes>::EquationIdVector(
```
Same remarks as for routine void TransientThermalElement<TDim, TNumNodes>::GetDofList

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+            rValues.resize(N_DOF, false);
+
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            rValues[i] = 0.0;
```
noalias( rValues ) = ZeroVector( N_DOF );

mnabideltares replied:
This function, with other 4 fuctions are found unnessary and removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+            rValues.resize(N_DOF, false);
+
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            rValues[i] = 0.0;
```
ZeroVector

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+            rValues.resize(N_DOF, false);
+
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            rValues[i] = 0.0;
```
ZeroVector

mnabideltares replied:
This function is removed

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+            KRATOS_ERROR << "DomainSize < 1.0e-15 for the element " << this->Id() << std::endl;
+
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            if (Geom[i].SolutionStepsDataHas(TEMPERATURE) == false)
```
if (!Geom[i].SolutionStepsDataHas(TEMPERATURE))

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+        if (TDim == 2) {
+            // If this is a 2D problem, nodes must be in XY plane
+            for (unsigned int i = 0; i < TNumNodes; ++i) {
+                if (Geom[i].Z() != 0.0)
```
std::any_of

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+        this->InitializeNodalTemperatureVariables(rVariables);
+
+        //Variables computed at each GP
+        rVariables.Np.resize(TNumNodes, false);
```
Np seems to refer to the polynomial for pressure. In a heat only element NT seems appropriate or even just N could be enough.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+
+        //Loop over integration points
+        for (unsigned int GPoint = 0; GPoint < NumGPoints; ++GPoint) {
+            //Compute GradNpT, B and StrainVector
```
GradNT or just GradN

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+    	                * rVariables.SolidHeatCapacity;
+        TMatrix = (c1 + c2) * outer_prod(rVariables.Np, rVariables.Np)
+    	                * rVariables.IntegrationCoefficient
+    	                * rVariables.DtTemperatureCoefficient;
```
suggested naming c1 --> cWater c2 --> cSolid

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+
+    	this->CalculateThermalDispersionMatrix(rVariables.ConstitutiveMatrix, rVariables);
+
+        BoundedMatrix<double, TDim, TNumNodes> Temp = prod(rVariables.ConstitutiveMatrix, trans(rVariables.GradNpT));
```
Temp here is likely temporary. In a temperature computation this may be confusing.

mnabideltares replied:
Temp is let as it is, it is meant to be temporary

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+        const double c2 = (1.0 - rVariables.Porosity) * rVariables.SolidDensity
+    	                * rVariables.SolidHeatCapacity;
+        TMatrix = (c1 + c2) * outer_prod(rVariables.Np, rVariables.Np)
+    	                * rVariables.IntegrationCoefficient;
```
This is a code repetition of the formulas in CalculateCapacityMatrix. Writing a function for these formulas avoids the repetition.

mnabideltares replied:
The repetition is removed. Now all in GeoThermalDispersion2DLaw

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+        C(0, 0) = c1 * rVariables.SolidThermalConductivityXX + c0;
+        C(0, 1) = c1 * rVariables.SolidThermalConductivityXY;
+        C(1, 0) = c1 * rVariables.SolidThermalConductivityYX;
+        C(1, 1) = c1 * rVariables.SolidThermalConductivityYY + c0;
```
Code repetition of void GeoThermalDispersion2DLaw::CalculateThermalDispersionMatrix(

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+
+        const bool CalculateStiffnessMatrixFlag = false;
+        const bool CalculateResidualVectorFlag = true;
+        MatrixType TempMatrix = Matrix();
```
The "StiffnessMatrix" is not computed, but used for the rRightHandSideVector. Is it necessary to make a copy, not computing it should not destroy what is there.

mnabideltares replied:
function for StiffnessMatrix is removed

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,763 @@
+    template< unsigned int TDim, unsigned int TNumNodes >
+    GeometryData::IntegrationMethod TransientThermalElement<TDim, TNumNodes>::GetIntegrationMethod() const
+    {
+        return GeometryData::IntegrationMethod::GI_GAUSS_2;
```
Insufficient for the higher order elements.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.hpp:
```diff
@@ -0,0 +1,220 @@
+            BoundedMatrix<double, TDim, TDim> ConstitutiveMatrix;
+            Vector Np;
+            Matrix GradNpT;
+            Matrix GradNpTInitialConfiguration;
```
Naming refers to pressure?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.hpp:
```diff
@@ -0,0 +1,220 @@
+    protected:
+
+        GeometryData::IntegrationMethod mThisIntegrationMethod;
+        /// Member Variables
```
comment before the member variable mThisIntegrationMethod?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.hpp:
```diff
@@ -0,0 +1,220 @@
+
+    private:
+
+        /// Member Variables
```
Nothing of member variables here

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.hpp:
```diff
@@ -0,0 +1,220 @@
+        TransientThermalElement& operator=(TransientThermalElement const& rOther);
+
+        // Copy constructor.
+        TransientThermalElement(TransientThermalElement const& rOther);
```
Private so not to be used. This also possible by declaration in the public part with = delete;

mnabideltares replied:
Removed, as this variable is never used

WPK4FEM commented on applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,154 @@
+    using LocalSystemMatrixType = typename BaseType::LocalSystemMatrixType;
+    using MotherType = NewmarkQuasistaticUPwScheme<TSparseSpace, TDenseSpace>;
+    using MotherType::mDeltaTime;
+    using MotherType::mBeta;
+    using MotherType::mGamma;
```
Never used here?

mnabideltares replied:
This problem is overal in the code, also for pressure.
I think it is not suitable to fix it in the thermal branch while the pressure then has to be solved sepsrately. 
I will make a new issue and address this issue (in a new branch).

WPK4FEM commented on applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,154 @@
+
+        //check for minimum value of the buffer index.
+        if (rModelPart.GetBufferSize() < 2)
+            KRATOS_ERROR << "insufficient buffer size. Buffer size should be greater than 2. Current size is "
```
error text inconsistent with condition. greater than or equal to 2

WPK4FEM commented on applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,154 @@
+                                        - rNode.FastGetSolutionStepValue(TEMPERATURE, 1);
+            const auto &PreviousDtTemperature = rNode.FastGetSolutionStepValue(DT_TEMPERATURE, 1);
+
+            rNode.FastGetSolutionStepValue(DT_TEMPERATURE) =  1.0/(mTheta*mDeltaTime)*(DeltaTemperature - (1.0-mTheta)*mDeltaTime*PreviousDtTemperature);
```
dimensional analysis raises doubt about this formula for mTheta other than 1.

mnabideltares replied:
This problem is overal in the code, also for pressure and deformation.
I think it is not suitable to fix it in the thermal branch. I will make a new issue and address this issue separately (in a new branch).

WPK4FEM commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -310,6 +328,20 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_VARIABLE( BULK_MODULUS_FLUID )
 
+    KRATOS_REGISTER_VARIABLE( HEAT_CAPACITY_WATER )
+    KRATOS_REGISTER_VARIABLE( HEAT_CAPACITY_SOLID )
```
This refers to specific heat capacity ( as opposed to volumetric heat capacity ). That could reflect in the naming.

mnabideltares replied:
SPECIFIC_HEAT_CAPACITY_WATER 
SPECIFIC_HEAT_CAPACITY_SOLID

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_T_solver.py:
```diff
@@ -0,0 +1,209 @@
+    return TSolver(model, custom_settings)
+
+class TSolver(GeoSolver):
+    '''Solver for the solution of displacement-pore pressure coupled problems.'''
```
I expect "Solver for the solution of temperature problems" or " heat problems"

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_T_solver.py:
```diff
@@ -0,0 +1,209 @@
+        if (scheme_type.lower() == "newmark" or scheme_type.lower() == "newmark_flow"):
+            theta = self.settings["newmark_theta"].GetDouble()
+            rayleigh_m = self.settings["rayleigh_m"].GetDouble()
+            rayleigh_k = self.settings["rayleigh_k"].GetDouble()
```
Not relevant, there is no damping matrix constructed from mass and stiffness matrices in a T solver

mnabideltares replied:
Removed

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py:
```diff
@@ -22,10 +22,16 @@ def CreateSolver(model, custom_settings):
         else:
             err_msg =  "The requested solver type \"" + solver_type + "\" is not in the python solvers wrapper\n"
-            err_msg += "Available options are: \"geomechanics_U_Pw_solver\", \"geomechanics_Pw_solver\""
+            err_msg += "Available options are: \"geomechanics_U_Pw_solver\", \"geomechanics_Pw_solver\", \"geomechanics_t_solver\""
```
capital T elsewhere

WPK4FEM reviewed (commented):
Hopefully consistent with our talk yesterday.

mnabideltares committed:
A bug fix

mnabideltares committed:
Removed a redundant variable

mnabideltares committed:
Another redundant variable is removed, mThisIntegrationMethod

mnabideltares committed:
A variable is removed from protected and defined it locally

mnabideltares committed:
Another redundant variable is removed

mnabideltares committed:
Modifications based on review, plus extra clean up

mnabideltares committed:
Fix a bug

mnabideltares committed:
fixed a typo bug

mnabideltares committed:
Addressed code smells

mnabideltares committed:
Fixed code smells

mnabideltares committed:
Modifications based on review process

avdg81 committed:
Fixed several code smells reported by SonarCloud

- Removed default implementations of the default constructor, the
  destructor, and the copy constructor.
- Explicitly captured the `this` pointer rather than capturing all
  variables in scope by reference.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+    Geom.Jacobian( JContainer, this->GetIntegrationMethod());
+    
+    //Condition variables
+    array_1d<double,TNumNodes> NormalFluxVector;
```
normal_flux_vector

mnabideltares replied:
@AlejandroCornejo  Thank you for the review. 
The following I addressed fixes based on your review.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+{
+    Vector NormalVector = ZeroVector(TDim);
+
+    if (TDim == 2)
```
you can use the newly added `if constexpr (TDim == 2){` which checks the` if` at compile time --> faster

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,642 @@
+    	const GeometryType& rGeom = this->GetGeometry();
+
+        //Nodal Variables
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
```
`IndexType`

mnabideltares replied:
what do you mean? We defined"
    using IndexType = std::size_t;
while TNumNodes is of type "unsigned int".
To be consistent, "unsigned int i" is applied here.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,642 @@
+        GeoThermalDispersion2DLaw::CalculateThermalDispersionMatrix(rVariables.ConstitutiveMatrix, rProp);
+
+        BoundedMatrix<double, TDim, TNumNodes> Temp = prod(rVariables.ConstitutiveMatrix, trans(rVariables.GradNT));
+        TMatrix = prod(rVariables.GradNT, Temp) * rVariables.IntegrationCoefficient;
```
noalias

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,642 @@
+        TMatrix = (cWater + cSolid) * outer_prod(rVariables.N, rVariables.N)
+            * rVariables.IntegrationCoefficient;
+
+        TVector = - prod(TMatrix, rVariables.DtTemperatureVector);
```
noalias

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,642 @@
+
+        TDimMatrix = prod(rVariables.GradNT, rVariables.ConstitutiveMatrix);
+
+        TMatrix = prod(TDimMatrix, trans(rVariables.GradNT))
```
noalias

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,642 @@
+    {
+        KRATOS_TRY
+
+        TDimMatrix = prod(rVariables.GradNT, rVariables.ConstitutiveMatrix);
```
noalias

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_strategies/schemes/backward_euler_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,72 @@
+    ///Destructor
+    // ============================================================================================
+    // ============================================================================================
+    ~BackwardEulerQuasistaticTScheme() override = default;
```
I thnk you can remove this

AlejandroCornejo commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_T_solver.py:
```diff
@@ -0,0 +1,203 @@
+            "block_builder"              : true,
+            "search_neighbours_step"     : false,
+            "linear_solver_settings":{
+                "solver_type": "AMGCL",
```
just for my curiosity, are you using by default AMGCL? does it performs well in your solid mechanics cases?

mnabideltares replied:
for pressure and deformation-pressure, this solver is applied. In order to be consistent with the other parts of the code, I kept it as AMGCL.

However, AMGCL has some advatages. The multigrid scales linearly with the number of nodes. However, Krylove space methods scale exponentially. Then for large number of nodes, multigrid is more efficient. from other side, for small number of nodes Krylove is faster. As we assume, we usually have large number of nodes, per default AMGCL is "probably" here set as the default solver.

AlejandroCornejo commented on applications/GeoMechanicsApplication/tests/test_thermal_fixed_temperature_2D3N.gid/MaterialParameters.json:
```diff
@@ -0,0 +1,28 @@
+              "DENSITY_WATER"            :  1.0,
+              "POROSITY"                 :  0.36,
+			  "SATURATION"               :  1.0,
+			  "SPECIFIC_HEAT_CAPACITY_SOLID"     :  400.0,
```
bad indentations here

AlejandroCornejo commented on applications/GeoMechanicsApplication/tests/test_thermal_fixed_temperature_2D6N.gid/MaterialParameters.json:
```diff
@@ -0,0 +1,28 @@
+              "DENSITY_WATER"            :  1.0,
+              "POROSITY"                 :  0.36,
+			  "SATURATION"               :  1.0,
+			  "SPECIFIC_HEAT_CAPACITY_SOLID"     :  400.0,
```
idem

AlejandroCornejo commented on applications/GeoMechanicsApplication/tests/test_thermal_heat_flux_2D6N.gid/MaterialParameters.json:
```diff
@@ -0,0 +1,28 @@
+			  "SATURATION"               :  1.0,
+			  "SPECIFIC_HEAT_CAPACITY_SOLID"     :  400.0,
+			  "SPECIFIC_HEAT_CAPACITY_WATER"     :  3795.0,
+			  "THERMAL_CONDUCTIVITY_SOLID_XX"    :  1.50,
```
.

mnabideltares committed:
Addressed the review requests

mnabideltares committed:
Merge branch 'master' into geo/10372-HeatTransfer-temp

# Conflicts:
#	applications/GeoMechanicsApplication/geo_mechanics_application.h
#	applications/GeoMechanicsApplication/geo_mechanics_application_constants.h

mnabideltares committed:
Merge branch 'master' into geo/10372-HeatTransfer-temp

mnabideltares committed:
Added README.md to serve the review process.

NB: This file must not be merged to the master

mnabideltares committed:
fix1

mnabideltares committed:
Readme.md is modified

mnabideltares committed:
Readme.md is modified further

mnabideltares committed:
fix in readme.md

mnabideltares committed:
Readme.md another fix

mnabideltares committed:
Readme.md fix

mnabideltares committed:
Thermal tests are extended to cover higher order elements.

NB: the lacking higher order elements and conditions are registered in the geo application

mnabideltares committed:
Added 6 more test casesd for thermal element,

mnabideltares committed:
Added the missing conditions of 3D, for thermal element

mnabideltares committed:
Changed some variable names to be lower case for the first letter, plus cleanup

mnabideltares committed:
Cleaned up the variables which are sent to functions, Removed repeatation of cod

introduced conductivity and capacity matrices and arrays to be used in later stages, to avoid double calculations

mnabideltares committed:
Included some of reviewer points from cosim branch, which are relavant to this branch.

mnabideltares committed:
Cleanup

mnabideltares committed:
Cleanup 2

mnabideltares committed:
Removed unused line to avoid WError (Warning Errors)

mnabideltares committed:
Test cases are fixed, the relative density is replaced by the physical value.

mnabideltares committed:
A new directory is opened to include all thermal elements together, and all moved to there

mnabideltares committed:
Input for 3D thermal element is added, and the constatutive law is modified to be 2D, 3D generic

mnabideltares committed:
Fixed a small bug

mnabideltares committed:
Swaping is removed for Triangle3D3N and Triangle3D6N. Swaping was done because of a bug, which was fixed. No swaping is needed more.

issues #9818  #9422 , PR #10541

mnabideltares committed:
4 extra test cases are added for 3D thermal element

mnabideltares committed:
unsteady thermal test cases are added

mnabideltares committed:
Cleanup in the Geo solver

mnabideltares committed:
Registered Newmark for the thermal element

mnabideltares committed:
Made the capacity and conductivity matrices as element based matrices (replacing of Gauss points base)
